File: ios/FamilyConnect/Models/EmojiCatalog.swift
//
//  EmojiCatalog.swift
//  FamilyConnect
//
//  The full-picker emoji catalog behind the floating picker's "+" sheet.
//  Transcribed from the canonical cross-platform catalog (the Android app
//  embeds the same list — same categories, same order — so the picker is
//  identical on both platforms; do not add, remove, or reorder entries
//  here without changing it there too). Every entry is a single-grapheme,
//  skin-tone-neutral emoji well under the server's 32-byte UTF-8 limit —
//  the unit tests pin all of that down.
//
//  One deliberate departure from the source JSON: its Hearts category
//  listed 💢 twice; the second occurrence is dropped here (a duplicate
//  offers the same reaction twice and collides ForEach identity). The
//  Android transcription must drop the same entry.
//
//  `name` is the canonical cross-platform identifier, not a UI string —
//  the section headers render `localizedName`.
//

import Foundation

/// One section of the full emoji picker: a canonical name and its emoji,
/// in canonical order.
nonisolated struct EmojiCategory: Equatable, Sendable, Identifiable {
    var id: String { name }
    /// Canonical category name, shared verbatim with Android.
    let name: String
    let emoji: [String]

    /// The localized section header for `name`. A literal per case so the
    /// string catalog's extractor sees every key.
    var localizedName: String {
        switch name {
        case "Smileys": String(localized: "Smileys")
        case "Gestures": String(localized: "Gestures")
        case "Hearts": String(localized: "Hearts")
        case "Animals & Nature": String(localized: "Animals & Nature")
        case "Food & Drink": String(localized: "Food & Drink")
        case "Activities": String(localized: "Activities")
        case "Travel & Places": String(localized: "Travel & Places")
        case "Objects & Symbols": String(localized: "Objects & Symbols")
        default: name
        }
    }
}

nonisolated enum EmojiCatalog {

    static let categories: [EmojiCategory] = [
        EmojiCategory(
            name: "Smileys",
            emoji: [
                "😀", "😃", "😄", "😁", "😆", "😅", "😂", "🤣", "🙂", "😉", "😊", "😇", "🥰", "😍", "🤩", "😘",
                "😗", "😚", "😙", "😋", "😛", "😜", "🤪", "😝", "🤑", "🤗", "🤭", "🤫", "🤔", "🤐", "🤨", "😐",
                "😑", "😶", "😏", "😒", "🙄", "😬", "🤥", "😌", "😔", "😪", "🤤", "😴", "😷", "🤒", "🤕", "🤢",
                "🤮", "🤧", "🥵", "🥶", "🥴", "😵", "🤯", "🤠", "🥳", "😎", "🤓", "🧐", "😕", "😟", "🙁", "😮",
                "😯", "😲", "😳", "🥺", "😦", "😧", "😨", "😰", "😥", "😢", "😭", "😱", "😖", "😣", "😞", "😓",
                "😩", "😫", "🥱", "😤", "😡", "😠", "🤬", "😈", "👿", "💀", "🤡", "👻", "👽", "🤖", "💩",
            ]),
        EmojiCategory(
            name: "Gestures",
            emoji: [
                "👋", "🤚", "🖐", "✋", "🖖", "👌", "🤏", "✌️", "🤞", "🤟", "🤘", "🤙", "👈", "👉", "👆", "🖕",
                "👇", "☝️", "👍", "👎", "✊", "👊", "🤛", "🤜", "👏", "🙌", "👐", "🤲", "🤝", "🙏", "✍️", "💅",
                "🤳", "💪", "🦾", "👂", "👃", "🧠", "👀", "👁", "👅", "👄", "💋",
            ]),
        EmojiCategory(
            name: "Hearts",
            emoji: [
                "❤️", "🧡", "💛", "💚", "💙", "💜", "🖤", "🤍", "🤎", "💔", "❣️", "💕", "💞", "💓", "💗", "💖",
                "💘", "💝", "💟", "♥️", "💌", "💤", "💢", "💥", "💦", "💨", "💫", "⭐", "🌟", "✨", "⚡", "🔥",
                "💯", "🎉", "🎊", "🎈", "🎂", "🎁", "🏆", "🥇", "🥈", "🥉", "🏅",
            ]),
        EmojiCategory(
            name: "Animals & Nature",
            emoji: [
                "🐶", "🐱", "🐭", "🐹", "🐰", "🦊", "🐻", "🐼", "🐨", "🐯", "🦁", "🐮", "🐷", "🐸", "🐵", "🙈",
                "🙉", "🙊", "🐔", "🐧", "🐦", "🐤", "🦆", "🦅", "🦉", "🦇", "🐺", "🐗", "🐴", "🦄", "🐝", "🐛",
                "🦋", "🐌", "🐞", "🐜", "🕷", "🐢", "🐍", "🦎", "🐙", "🦑", "🦀", "🐡", "🐠", "🐟", "🐬", "🐳",
                "🐋", "🦈", "🐊", "🐘", "🦏", "🐪", "🦒", "🦘", "🐃", "🐎", "🐖", "🐏", "🐑", "🦙", "🐐", "🦌",
                "🐕", "🐩", "🐈", "🐇", "🐿", "🦔", "🌵", "🎄", "🌲", "🌳", "🌴", "🌱", "🌿", "☘️", "🍀", "🎍",
                "🍁", "🍄", "🌾", "💐", "🌷", "🌹", "🥀", "🌺", "🌸", "🌼", "🌻", "🌞", "🌝", "🌚", "🌙", "🌎",
                "🪐", "💫", "🌈", "☀️", "⛅", "☁️", "🌧", "⛈", "❄️", "⛄", "🌊",
            ]),
        EmojiCategory(
            name: "Food & Drink",
            emoji: [
                "🍏", "🍎", "🍐", "🍊", "🍋", "🍌", "🍉", "🍇", "🍓", "🍈", "🍒", "🍑", "🥭", "🍍", "🥥", "🥝",
                "🍅", "🍆", "🥑", "🥦", "🥬", "🥒", "🌶", "🌽", "🥕", "🧄", "🧅", "🥔", "🍠", "🥐", "🥯", "🍞",
                "🥖", "🥨", "🧀", "🥚", "🍳", "🧈", "🥞", "🧇", "🥓", "🥩", "🍗", "🍖", "🌭", "🍔", "🍟", "🍕",
                "🥪", "🥙", "🧆", "🌮", "🌯", "🥗", "🥘", "🍝", "🍜", "🍲", "🍛", "🍣", "🍱", "🥟", "🦪", "🍤",
                "🍙", "🍚", "🍘", "🍥", "🥮", "🍢", "🍡", "🍧", "🍨", "🍦", "🥧", "🧁", "🍰", "🎂", "🍮", "🍭",
                "🍬", "🍫", "🍿", "🍩", "🍪", "🌰", "🥜", "🍯", "🥛", "🍼", "☕", "🍵", "🧃", "🥤", "🍶", "🍺",
                "🍻", "🥂", "🍷", "🥃", "🍸", "🍹", "🧉", "🍾", "🧊",
            ]),
        EmojiCategory(
            name: "Activities",
            emoji: [
                "⚽", "🏀", "🏈", "⚾", "🥎", "🎾", "🏐", "🏉", "🥏", "🎱", "🪀", "🏓", "🏸", "🏒", "🏑", "🥍",
                "🏏", "🥅", "⛳", "🪁", "🏹", "🎣", "🤿", "🥊", "🥋", "🎽", "🛹", "🛷", "⛸", "🥌", "🎿", "⛷",
                "🏂", "🏋️", "🤸", "⛹️", "🤺", "🤾", "🏌️", "🏇", "🧘", "🏄", "🏊", "🤽", "🚣", "🧗", "🚴", "🚵",
                "🎪", "🎭", "🎨", "🎬", "🎤", "🎧", "🎼", "🎹", "🥁", "🎷", "🎺", "🎸", "🪕", "🎻", "🎲", "♟",
                "🎯", "🎳", "🎮", "🎰", "🧩",
            ]),
        EmojiCategory(
            name: "Travel & Places",
            emoji: [
                "🚗", "🚕", "🚙", "🚌", "🚎", "🏎", "🚓", "🚑", "🚒", "🚐", "🚚", "🚛", "🚜", "🛴", "🚲", "🛵",
                "🏍", "🚨", "🚔", "🚍", "🚘", "🚖", "🚡", "🚠", "🚟", "🚃", "🚋", "🚞", "🚝", "🚄", "🚅", "🚈",
                "🚂", "🚆", "🚇", "🚊", "🚉", "✈️", "🛫", "🛬", "🛩", "💺", "🛰", "🚀", "🛸", "🚁", "🛶", "⛵",
                "🚤", "🛥", "🛳", "⛴", "🚢", "⚓", "⛽", "🚧", "🚦", "🚥", "🚏", "🗺", "🗿", "🗽", "🗼", "🏰",
                "🏯", "🏟", "🎡", "🎢", "🎠", "⛲", "⛱", "🏖", "🏝", "🏜", "🌋", "⛰", "🏔", "🗻", "🏕", "⛺",
                "🏠", "🏡", "🏘", "🏚", "🏗", "🏭", "🏢", "🏬", "🏣", "🏤", "🏥", "🏦", "🏨", "🏪", "🏫", "🏩",
                "💒", "🏛", "⛪", "🕌", "🕍", "🛕", "🕋", "⛩", "🌅", "🌄", "🌠", "🎇", "🎆", "🌇", "🌆", "🏙",
                "🌃", "🌌", "🌉", "🌁",
            ]),
        EmojiCategory(
            name: "Objects & Symbols",
            emoji: [
                "⌚", "📱", "💻", "⌨️", "🖥", "🖨", "🖱", "🕹", "🗜", "💽", "💾", "💿", "📀", "📼", "📷", "📸",
                "📹", "🎥", "📽", "🎞", "📞", "☎️", "📟", "📠", "📺", "📻", "🎙", "⏰", "⌛", "⏳", "📡", "🔋",
                "🔌", "💡", "🔦", "🕯", "🧯", "🛢", "💸", "💵", "💰", "💳", "💎", "⚖️", "🧰", "🔧", "🔨", "⚒",
                "🛠", "⛏", "🔩", "⚙️", "🧲", "🔫", "💣", "🧨", "🪓", "🔪", "🗡", "🛡", "🚬", "⚰️", "⚱️", "🏺",
                "🔮", "📿", "🧿", "💈", "⚗️", "🔭", "🔬", "🕳", "💊", "💉", "🧬", "🦠", "🧫", "🧪", "🌡", "🧹",
                "🧺", "🧻", "🚽", "🚰", "🚿", "🛁", "🧼", "🪒", "🧽", "🧴", "🛎", "🔑", "🗝", "🚪", "🪑", "🛋",
                "🛏", "🧸", "🖼", "🛍", "🛒", "🎀", "🎏", "🎗", "📯", "📦", "📫", "📮", "📜", "📃", "📄", "📊",
                "📈", "📉", "🗒", "📆", "📅", "📇", "🗃", "🗳", "🗄", "📋", "📁", "📂", "🗞", "📰", "📓", "📔",
                "📒", "📕", "📗", "📘", "📙", "📚", "📖", "🔖", "🧷", "🔗", "📎", "📐", "📏", "🧮", "📌", "📍",
                "✂️", "🖊", "🖋", "✒️", "🖌", "🖍", "📝", "✏️", "🔍", "🔎", "🔒", "🔓", "❗", "❓", "‼️", "⁉️",
                "✅", "❌", "⭕", "🚫", "💬", "💭", "🗯", "♻️", "🔱", "📣", "📢", "🔔", "🔕", "🎵", "🎶", "➕",
                "➖", "➗", "✖️", "♾", "💲", "™️", "©️", "®️", "🔴", "🟠", "🟡", "🟢", "🔵", "🟣", "⚫", "⚪",
                "🟤",
            ]),
    ]
}

File: android/app/src/main/java/me/nettrash/familyconnect/ui/chat/EmojiCatalog.kt
/*
 * EmojiCatalog.kt
 * Family Connect (Android)
 *
 * The canonical full-picker emoji catalog behind the reaction capsule's
 * "+" button. Transcribed from the shared cross-platform catalog — the
 * iOS app embeds the IDENTICAL list (same categories, same order) so the
 * "+" picker looks the same on both platforms. Do not add, remove, or
 * reorder entries here without changing iOS in the same breath.
 *
 * Only widely-supported, skin-tone-neutral, single-grapheme emoji
 * (renders on Android 8 / iOS 17); every entry is well under the
 * server's 32-byte UTF-8 reaction limit — EmojiCatalogTest pins that.
 *
 * Kept free of Compose and Android so the sanity test runs as a plain
 * unit test. Category names are UI strings (the app's UI copy is inline
 * English throughout — see ChatScreen.kt).
 *
 * iOS counterpart: ios/FamilyConnect/Views/ConversationView.swift
 */

package me.nettrash.familyconnect.ui.chat

import androidx.annotation.StringRes
import me.nettrash.familyconnect.R

/** One section of the full emoji picker. */
data class EmojiCategory(
    /** The category's stable English name — the grid's key and the tests' handle, never drawn. */
    val name: String,
    /** The section header, through the catalogue (nine languages). */
    @param:StringRes val titleRes: Int,
    val emoji: List<String>,
)

/** Every category the "+" picker offers, in display order. */
val EMOJI_CATALOG: List<EmojiCategory> = listOf(
    EmojiCategory(
        name = "Smileys",
        titleRes = R.string.s_emoji_smileys,
        emoji = listOf(
            "😀", "😃", "😄", "😁", "😆", "😅", "😂", "🤣", "🙂", "😉", "😊", "😇",
            "🥰", "😍", "🤩", "😘", "😗", "😚", "😙", "😋", "😛", "😜", "🤪", "😝",
            "🤑", "🤗", "🤭", "🤫", "🤔", "🤐", "🤨", "😐", "😑", "😶", "😏", "😒",
            "🙄", "😬", "🤥", "😌", "😔", "😪", "🤤", "😴", "😷", "🤒", "🤕", "🤢",
            "🤮", "🤧", "🥵", "🥶", "🥴", "😵", "🤯", "🤠", "🥳", "😎", "🤓", "🧐",
            "😕", "😟", "🙁", "😮", "😯", "😲", "😳", "🥺", "😦", "😧", "😨", "😰",
            "😥", "😢", "😭", "😱", "😖", "😣", "😞", "😓", "😩", "😫", "🥱", "😤",
            "😡", "😠", "🤬", "😈", "👿", "💀", "🤡", "👻", "👽", "🤖", "💩",
        ),
    ),
    EmojiCategory(
        name = "Gestures",
        titleRes = R.string.s_emoji_gestures,
        emoji = listOf(
            "👋", "🤚", "🖐", "✋", "🖖", "👌", "🤏", "✌️", "🤞", "🤟", "🤘", "🤙",
            "👈", "👉", "👆", "🖕", "👇", "☝️", "👍", "👎", "✊", "👊", "🤛", "🤜",
            "👏", "🙌", "👐", "🤲", "🤝", "🙏", "✍️", "💅", "🤳", "💪", "🦾", "👂",
            "👃", "🧠", "👀", "👁", "👅", "👄", "💋",
        ),
    ),
    EmojiCategory(
        name = "Hearts",
        titleRes = R.string.s_emoji_hearts,
        emoji = listOf(
            "❤️", "🧡", "💛", "💚", "💙", "💜", "🖤", "🤍", "🤎", "💔", "❣️", "💕",
            "💞", "💓", "💗", "💖", "💘", "💝", "💟", "♥️", "💌", "💤", "💢", "💥",
            "💦", "💨", "💫", "⭐", "🌟", "✨", "⚡", "🔥", "💯", "🎉", "🎊", "🎈",
            "🎂", "🎁", "🏆", "🥇", "🥈", "🥉", "🏅",
        ),
    ),
    EmojiCategory(
        name = "Animals & Nature",
        titleRes = R.string.s_emoji_animals,
        emoji = listOf(
            "🐶", "🐱", "🐭", "🐹", "🐰", "🦊", "🐻", "🐼", "🐨", "🐯", "🦁", "🐮",
            "🐷", "🐸", "🐵", "🙈", "🙉", "🙊", "🐔", "🐧", "🐦", "🐤", "🦆", "🦅",
            "🦉", "🦇", "🐺", "🐗", "🐴", "🦄", "🐝", "🐛", "🦋", "🐌", "🐞", "🐜",
            "🕷", "🐢", "🐍", "🦎", "🐙", "🦑", "🦀", "🐡", "🐠", "🐟", "🐬", "🐳",
            "🐋", "🦈", "🐊", "🐘", "🦏", "🐪", "🦒", "🦘", "🐃", "🐎", "🐖", "🐏",
            "🐑", "🦙", "🐐", "🦌", "🐕", "🐩", "🐈", "🐇", "🐿", "🦔", "🌵", "🎄",
            "🌲", "🌳", "🌴", "🌱", "🌿", "☘️", "🍀", "🎍", "🍁", "🍄", "🌾", "💐",
            "🌷", "🌹", "🥀", "🌺", "🌸", "🌼", "🌻", "🌞", "🌝", "🌚", "🌙", "🌎",
            "🪐", "💫", "🌈", "☀️", "⛅", "☁️", "🌧", "⛈", "❄️", "⛄", "🌊",
        ),
    ),
    EmojiCategory(
        name = "Food & Drink",
        titleRes = R.string.s_emoji_food,
        emoji = listOf(
            "🍏", "🍎", "🍐", "🍊", "🍋", "🍌", "🍉", "🍇", "🍓", "🍈", "🍒", "🍑",
            "🥭", "🍍", "🥥", "🥝", "🍅", "🍆", "🥑", "🥦", "🥬", "🥒", "🌶", "🌽",
            "🥕", "🧄", "🧅", "🥔", "🍠", "🥐", "🥯", "🍞", "🥖", "🥨", "🧀", "🥚",
            "🍳", "🧈", "🥞", "🧇", "🥓", "🥩", "🍗", "🍖", "🌭", "🍔", "🍟", "🍕",
            "🥪", "🥙", "🧆", "🌮", "🌯", "🥗", "🥘", "🍝", "🍜", "🍲", "🍛", "🍣",
            "🍱", "🥟", "🦪", "🍤", "🍙", "🍚", "🍘", "🍥", "🥮", "🍢", "🍡", "🍧",
            "🍨", "🍦", "🥧", "🧁", "🍰", "🎂", "🍮", "🍭", "🍬", "🍫", "🍿", "🍩",
            "🍪", "🌰", "🥜", "🍯", "🥛", "🍼", "☕", "🍵", "🧃", "🥤", "🍶", "🍺",
            "🍻", "🥂", "🍷", "🥃", "🍸", "🍹", "🧉", "🍾", "🧊",
        ),
    ),
    EmojiCategory(
        name = "Activities",
        titleRes = R.string.s_emoji_activities,
        emoji = listOf(
            "⚽", "🏀", "🏈", "⚾", "🥎", "🎾", "🏐", "🏉", "🥏", "🎱", "🪀", "🏓",
            "🏸", "🏒", "🏑", "🥍", "🏏", "🥅", "⛳", "🪁", "🏹", "🎣", "🤿", "🥊",
            "🥋", "🎽", "🛹", "🛷", "⛸", "🥌", "🎿", "⛷", "🏂", "🏋️", "🤸", "⛹️",
            "🤺", "🤾", "🏌️", "🏇", "🧘", "🏄", "🏊", "🤽", "🚣", "🧗", "🚴", "🚵",
            "🎪", "🎭", "🎨", "🎬", "🎤", "🎧", "🎼", "🎹", "🥁", "🎷", "🎺", "🎸",
            "🪕", "🎻", "🎲", "♟", "🎯", "🎳", "🎮", "🎰", "🧩",
        ),
    ),
    EmojiCategory(
        name = "Travel & Places",
        titleRes = R.string.s_emoji_travel,
        emoji = listOf(
            "🚗", "🚕", "🚙", "🚌", "🚎", "🏎", "🚓", "🚑", "🚒", "🚐", "🚚", "🚛",
            "🚜", "🛴", "🚲", "🛵", "🏍", "🚨", "🚔", "🚍", "🚘", "🚖", "🚡", "🚠",
            "🚟", "🚃", "🚋", "🚞", "🚝", "🚄", "🚅", "🚈", "🚂", "🚆", "🚇", "🚊",
            "🚉", "✈️", "🛫", "🛬", "🛩", "💺", "🛰", "🚀", "🛸", "🚁", "🛶", "⛵",
            "🚤", "🛥", "🛳", "⛴", "🚢", "⚓", "⛽", "🚧", "🚦", "🚥", "🚏", "🗺",
            "🗿", "🗽", "🗼", "🏰", "🏯", "🏟", "🎡", "🎢", "🎠", "⛲", "⛱", "🏖",
            "🏝", "🏜", "🌋", "⛰", "🏔", "🗻", "🏕", "⛺", "🏠", "🏡", "🏘", "🏚",
            "🏗", "🏭", "🏢", "🏬", "🏣", "🏤", "🏥", "🏦", "🏨", "🏪", "🏫", "🏩",
            "💒", "🏛", "⛪", "🕌", "🕍", "🛕", "🕋", "⛩", "🌅", "🌄", "🌠", "🎇",
            "🎆", "🌇", "🌆", "🏙", "🌃", "🌌", "🌉", "🌁",
        ),
    ),
    EmojiCategory(
        name = "Objects & Symbols",
        titleRes = R.string.s_emoji_objects,
        emoji = listOf(
            "⌚", "📱", "💻", "⌨️", "🖥", "🖨", "🖱", "🕹", "🗜", "💽", "💾", "💿",
            "📀", "📼", "📷", "📸", "📹", "🎥", "📽", "🎞", "📞", "☎️", "📟", "📠",
            "📺", "📻", "🎙", "⏰", "⌛", "⏳", "📡", "🔋", "🔌", "💡", "🔦", "🕯",
            "🧯", "🛢", "💸", "💵", "💰", "💳", "💎", "⚖️", "🧰", "🔧", "🔨", "⚒",
            "🛠", "⛏", "🔩", "⚙️", "🧲", "🔫", "💣", "🧨", "🪓", "🔪", "🗡", "🛡",
            "🚬", "⚰️", "⚱️", "🏺", "🔮", "📿", "🧿", "💈", "⚗️", "🔭", "🔬", "🕳",
            "💊", "💉", "🧬", "🦠", "🧫", "🧪", "🌡", "🧹", "🧺", "🧻", "🚽", "🚰",
            "🚿", "🛁", "🧼", "🪒", "🧽", "🧴", "🛎", "🔑", "🗝", "🚪", "🪑", "🛋",
            "🛏", "🧸", "🖼", "🛍", "🛒", "🎀", "🎏", "🎗", "📯", "📦", "📫", "📮",
            "📜", "📃", "📄", "📊", "📈", "📉", "🗒", "📆", "📅", "📇", "🗃", "🗳",
            "🗄", "📋", "📁", "📂", "🗞", "📰", "📓", "📔", "📒", "📕", "📗", "📘",
            "📙", "📚", "📖", "🔖", "🧷", "🔗", "📎", "📐", "📏", "🧮", "📌", "📍",
            "✂️", "🖊", "🖋", "✒️", "🖌", "🖍", "📝", "✏️", "🔍", "🔎", "🔒", "🔓",
            "❗", "❓", "‼️", "⁉️", "✅", "❌", "⭕", "🚫", "💬", "💭", "🗯", "♻️",
            "🔱", "📣", "📢", "🔔", "🔕", "🎵", "🎶", "➕", "➖", "➗", "✖️", "♾",
            "💲", "™️", "©️", "®️", "🔴", "🟠", "🟡", "🟢", "🔵", "🟣", "⚫", "⚪",
            "🟤",
        ),
    ),
)
